docs/RFCS/replica_tombstone.md

```diff
@@ -0,0 +1,113 @@
+generated using a `next_replica_id` field which will be added to the
+`RangeDescriptor` (alternative generation strategies include using the
+raft log position or database timestamp of the `EndTransaction` call
+that commits the membership change, but this information is less
+accessible at the point where it would be needed).
+
+The `ReplicaID` replaces the current `RaftNodeID` (which is
+constructed from the `NodeID` and `StoreID`). Raft only knows about
+replica IDs, which are never reused, so we don't have to worry about
+certain problems that can aries when node IDs are reused (such as vote
+changes or log regression. `MultiRaft` is responsible for mapping
+`ReplicaIDs` to node and store IDs (which it must do to coalesce
+heartbeats). This is done with a new method on the
+`WriteableGroupStorage` interface, along with an in-memory cache. Note
+the node and store IDs for a given replica never change once that
+replica is created, so we don't need to worry about synchronizing or
+invalidating entries in this cache.
+
+The raft transport will send the node, store, and replica ID of both
+the sender and receiver with every `RaftMessageRequest`. This is how
+the `Store` will learn of the replica ID for new ranges (the sender's
+IDs must be inserted in the replica ID cache, so we can route
+responses that do not yet appear in any range descriptor we have
+stored). The `Store` will drop incoming messages with a replica ID
+that is less than the last known one for the range (in order to
+minimize disruption from out-of-date servers).
+
+The `DistSender` will also send the replica ID in the header of every
+request, and the receiver will reject requests with incorrect replica
+IDs (This will be rare since a range will normally be absent from a
+node for a time before being re-replicated).
+
+When a replica is garbage-collected, we write a **tombstone**
+containing the replica ID, so that the store can continue to drop
+out-of-date raft messages even after the GC.
+
+In the scenario above, with this change node C would send node A and
+B's replica IDs in its vote requests. They would see that they have a
+tombstone for that replica and not recreate it.
+
+# Drawbacks
+
+Tombstones must be long-lived, since a node may come back online after
+a lengthy delay. We cannot garbage-collect tombstones unless we also
+guarantee that no node will come back after a period of time longer
+than the tombstone GC time.
+
+# Alternatives
+
+If we had an explicit replica-creation RPC instead of creating them
+automatically on the first sighting of a raft message, this problem
+may go away. However, doing so would be tricky: this RPC would need to
+be retried in certain circumstances, and it is difficult to
+distinguish cases where a retry is necessary from cases that lead to
+the problem discussed here.
+
+# Unresolved questions
+
+* How long must we wait before garbage-collecting tombstones? Can we
```

Comment:
I'm not sure this question needs to be resolved. Let's not cache the tombstones. It should be rare to get a message for which there's no legitimate range. Either we're creating a new one in the normal course of rebalancing or repair, or it's an old message--both should be exceptions compared to the normal flow of traffic. Since we're able to go to disk, the total number of tombstones can be incredibly large without causing much of a headache.